Changes in module 'nf-core/bedtools/intersect'
--- modules/nf-core/bedtools/intersect/main.nf
+++ modules/nf-core/bedtools/intersect/main.nf
@@ -9,7 +9,7 @@
 
     input:
     tuple val(meta), path(intervals1), path(intervals2)
-    tuple val(meta2), path(chrom_sizes)
+    val extension
 
     output:
     tuple val(meta), path("*.${extension}"), emit: intersect
@@ -22,8 +22,6 @@
     def args = task.ext.args ?: ''
     def prefix = task.ext.prefix ?: "${meta.id}"
     //Extension of the output file. It is set by the user via "ext.suffix" in the config. Corresponds to the file format which depends on arguments (e. g., ".bed", ".bam", ".txt", etc.).
-    extension = task.ext.suffix ?: "bed"
-    def sizes = chrom_sizes ? "-g ${chrom_sizes}" : ''
     if ("$intervals1" == "${prefix}.${extension}" ||
         "$intervals2" == "${prefix}.${extension}")
         error "Input and output names are the same, use \"task.ext.prefix\" to disambiguate!"
@@ -33,7 +31,6 @@
         -a $intervals1 \\
         -b $intervals2 \\
         $args \\
-        $sizes \\
         > ${prefix}.${extension}
 
     cat <<-END_VERSIONS > versions.yml
@@ -44,7 +41,6 @@
 
     stub:
     def prefix = task.ext.prefix ?: "${meta.id}"
-    extension = task.ext.suffix ?: "bed"
     if ("$intervals1" == "${prefix}.${extension}" ||
         "$intervals2" == "${prefix}.${extension}")
         error "Input and output names are the same, use \"task.ext.prefix\" to disambiguate!"

************************************************************
